src/hotspot/share/classfile/resolutionErrors.hpp

```diff
@@ -89,46 +63,65 @@ class ResolutionErrorTable : public Hashtable<ConstantPool*, mtClass> {
     assert(index < 0, "Unexpected non-negative cpCache index");
     return index + CPCACHE_INDEX_MANGLE_VALUE;
   }
+
+  static uintptr_t convert_key(const constantPoolHandle& pool, int cp_index) {
+    return (uintptr_t) (pool() + cp_index);
+  }
 };
 
 
-class ResolutionErrorEntry : public HashtableEntry<ConstantPool*, mtClass> {
+class ResolutionErrorEntry : public CHeapObj<mtClass> {
  private:
   int               _cp_index;
   Symbol*           _error;
   Symbol*           _message;
   Symbol*           _cause;
   Symbol*           _cause_msg;
+  ConstantPool*     _pool;
   const char*       _nest_host_error;
 
  public:
-  ConstantPool*      pool() const               { return literal(); }
 
+    ResolutionErrorEntry(ConstantPool* pool, int cp_index, Symbol* error, Symbol* message,
+      Symbol* cause, Symbol* cause_msg):
+        _cp_index(cp_index),
+        _error(error),
+        _message(message),
+        _cause(cause),
+        _cause_msg(cause_msg),
+        _pool(pool),
+        _nest_host_error(nullptr) {
+
+      _error->increment_refcount();
+      _message->increment_refcount();
+      _cause->increment_refcount();
+      _cause_msg->increment_refcount();
+    }
```

Comment:
So I was thinking that you might want to move this to the .cpp file.  It requires symbol.hpp to be included - transitively through constantPool.hpp - and doesn't need to be performant.